### PR TITLE
plugins: qmake

### PR DESCRIPTION
**Pull request by ScarlettGatelyMoore**
- [x] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?

-----

I ran the make tests - but I am going to be honest and couldn't get past the ruff test. It wanted me to change a bunch of files that don't belong to me.

### CODE REVIEW

File: craft_parts/plugins/qmake_plugin.py
Comment (lengau) [inline craft_parts/plugins/qmake_plugin.py:85]: What happens currently if someone tries to build a qt6-only project with this plugin? (I'm fine with it failing, I just want to ensure it fails "nicely".)

File: tests/unit/plugins/test_qmake_plugin.py
Comment (lengau) [inline tests/unit/plugins/test_qmake_plugin.py:3]:
```suggestion
# Copyright 2023 Canonical Ltd.
```

Feel free to add yourself or your org to the copyright block too (I'm guessing the copyright is partially Canonical's since I presume that more than the copyright block was copied from elsewhere - correct as needed :-) ) And same note about the copyright owners on the plugin file too.

**Review by lengau — COMMENTED** [review]
Thanks @ScarlettGatelyMoore, this looks great! I'd like @cmatsuoka to review this too since he's much more familiar with craft_parts than I am, but all my comments are pretty minor/nitpicky.

I've also enabled the CI runners here for tests and stuff.

**Comment by lengau** [comment]
@ScarlettGatelyMoore I'll take a look at the ruff stuff - it's probably due to problems with the configuration I made + ruff not being stable yet

File: tests/unit/plugins/test_qmake_plugin.py
Comment (ScarlettGatelyMoore) [reply tests/unit/plugins/test_qmake_plugin.py:3]: That is correct :) I followed a commit for the cmake plugin and converted to qmake logic. I will add myself.

**Comment by ScarlettGatelyMoore** [comment]
> 

Great, Thanks!

File: craft_parts/plugins/qmake_plugin.py
Comment (lengau) [reply craft_parts/plugins/qmake_plugin.py:86]: I guess what I'm really just looking for is that the error message be something along the lines of "the qmake plugin doesn't currently support qt6" rather than "error on line 4637453 of some_obscure_file.hpp: Missing }". As long as it's fairly clear to a user I'm not too concerned.

Comment (ScarlettGatelyMoore) [reply craft_parts/plugins/qmake_plugin.py]: Instead I added support for qmake6!

**Comment by ScarlettGatelyMoore** [comment]
Who do I need to poke to get these workflows approved?

**Comment by ScarlettGatelyMoore** [comment]
> I made a PR to your branch that fixes the linting issues on my machine (and presumably in CI): [ScarlettGatelyMoore#1](https://github.com/ScarlettGatelyMoore/craft-parts/pull/1)
> 
> I think it'll fix the pytest import error too

Hi, I am having issues trying to fix my tests as I still can't make tests locally due ruff errors on files I don't own. The tests on here require maintainer approval and hasn't had approval for days. Please help.

**Comment by capecrusader-121** [comment]
@lengau  @ScarlettGatelyMoore  Any progress on this or any targeted release ?

**Comment by ScarlettGatelyMoore** [comment]
> @lengau @ScarlettGatelyMoore Any progress on this or any targeted release ?

I got stuck and then time got away from me. I will take a look again this week. Or if another set of eyes can take a look, I don't oppose help.

**Comment by ScarlettGatelyMoore** [comment]
I made progress on this. The tests pass now except pyright fails on a bunch of __doc__ type failures and are unrelated to this.

File: craft_parts/plugins/qmake_plugin.py
Comment (cmatsuoka) [inline craft_parts/plugins/qmake_plugin.py:128]: Just curious, what does this command-line parameter mean?

File: craft_parts/plugins/qmake_plugin.py
Comment (cmatsuoka) [inline craft_parts/plugins/qmake_plugin.py:133]: Since this argument is the same in both cases, maybe we could add it before the conditional?

File: craft_parts/plugins/qmake_plugin.py
Comment (cmatsuoka) [inline craft_parts/plugins/qmake_plugin.py:7]: Craft-parts plugins should be LGPL.

File: craft_parts/plugins/qmake_plugin.py
Comment (ScarlettGatelyMoore) [reply craft_parts/plugins/qmake_plugin.py:128]: Sorry, not clear what you are asking here.  I have an optional qmake_project_file: str = "" property so that one can define a non standard .pro file as demonstrated in the test 


    def test_get_build_commands_qmake_project_file(self, setup_method_fixture, new_dir):
        plugin = setup_method_fixture(
            new_dir, properties={"qmake_project_file": "hello.pro"}

Comment (cmatsuoka) [reply craft_parts/plugins/qmake_plugin.py:130]: I asked because this list seems to evaluate to three different command-line arguments (src dir, /, and the project file) and I was wondering what the / could mean here. Or should they assemble a single path such as `srcdir/hello.pro`?

Comment (ScarlettGatelyMoore) [reply craft_parts/plugins/qmake_plugin.py:129]: Correct, assemble a single path

Comment (cmatsuoka) [reply craft_parts/plugins/qmake_plugin.py:129]: But then, when joining with a space in line 137 below, wouldn't that result in three different arguments?

Comment (cmatsuoka) [reply craft_parts/plugins/qmake_plugin.py:129]: One way to address this could be something like:
```
qmake_configure_command.append(str(self._part_info.part_src_dir / options.qmake_project_file))
```
in line 126 so that this specific entry would be a single argument.

Comment (ScarlettGatelyMoore) [reply craft_parts/plugins/qmake_plugin.py:129]: Thank you, a big problem is I can't seem to run make tests locally, ruff tests fail with things way beyond my level of knowledge. Is there a way to only run unit tests?

Comment (cmatsuoka) [reply craft_parts/plugins/qmake_plugin.py]: You can run `make test-units` or `make test-integrations` to run all of each kind or `pytest -vv tests/some_file.py` to run a single test.

**Comment by lengau** [comment]
I've made [a PR to your branch](https://github.com/ScarlettGatelyMoore/craft-parts/pull/2) that should fix the relevant issues.

**Review by lengau — APPROVED** [review]
LGTM! Thanks for all the work on this!

**Review by cmatsuoka — APPROVED** [review]
Thanks Scarlett and Alex! L.et's land this on main and test it -- if everything looks fine we can backport it to 1.x

**Pull request merged**
